File: backend/app/models/audit_log.py
from typing import Any, Dict, Optional

from sqlalchemy import JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.models.mixins import IDTimestampMixin


class AuditLog(IDTimestampMixin, Base):
    __tablename__ = "audit_logs"

    actor: Mapped[str] = mapped_column(String(32), nullable=False)
    action_type: Mapped[str] = mapped_column(String(120), nullable=False)
    target_type: Mapped[str] = mapped_column(String(120), nullable=False)
    target_id: Mapped[str] = mapped_column(String(36), nullable=False)
    before_snapshot: Mapped[Optional[Dict[str, Any]]] = mapped_column(JSON)
    after_snapshot: Mapped[Optional[Dict[str, Any]]] = mapped_column(JSON)
    note: Mapped[Optional[str]] = mapped_column(Text)

File: frontend/Sources/LinoFinanceCore/Models/AuditLog.swift
import Foundation

public struct AuditLog: Codable, Equatable, Sendable {
    public let id: String
    public let actor: String
    public let actionType: String
    public let targetType: String
    public let targetID: String
    public let beforeSnapshot: [String: JSONValue]?
    public let afterSnapshot: [String: JSONValue]?
    public let note: String?

    public init(
        id: String,
        actor: String,
        actionType: String,
        targetType: String,
        targetID: String,
        beforeSnapshot: [String: JSONValue]? = nil,
        afterSnapshot: [String: JSONValue]? = nil,
        note: String? = nil
    ) {
        self.id = id
        self.actor = actor
        self.actionType = actionType
        self.targetType = targetType
        self.targetID = targetID
        self.beforeSnapshot = beforeSnapshot
        self.afterSnapshot = afterSnapshot
        self.note = note
    }
}
